File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
    
//         ArrayList<Integer> [] maps = new ArrayList [n+1] ;
        
//         for (int i =0; i <n+1 ; i++) {
//             maps[i] = new ArrayList<>();
//         }
        
//         for (int [] road : roads)  {
//             int a = road[0];
//             int b = road [1];
//             maps[a].add(b);
//             maps[b].add(a);
//         }
        
//         ArrayDeque <Integer> q = new ArrayDeque<>();
        
//         // destination -> 1 까지 bfs
//         q.offer(destination);
        
//         int [] distance = new int [n+1];
        
//         Arrays.fill(distance, -1);
//         distance [destination] = 0;
//         while (!q.isEmpty()) {
//             int cur = q.poll();
            
//             for(int nxt : maps[cur]) {
//                 if (distance[nxt] == -1)  {
//                     distance[nxt] = distance[cur] +1;
//                     q.offer (nxt);
//                 }
//             }
            
//         }
        
//         int [] answer = new int [sources.length];
//         int idx= 0;
        
//         // 답 저장하기
//         for (int target : sources) {
//             answer[idx] = distance[target];
//             idx = idx +1;
//         }
//         return answer;
   
        ArrayList<Integer> [] maps = new ArrayList[n+1];
        
        for (int i =0; i<= n; i ++) {
            maps[i] = new ArrayList<>();
            
        }
        
        for (int[] r : roads) {
            int start = r[0];
            int end = r[1];
            maps[start].add(end);
            maps[end].add(start);
        }
        
        
        ArrayDeque <Integer> q = new ArrayDeque<>();
        q.offer(destination);
        
        int distance[] = new int[n+1];
        Arrays.fill(distance, -1);
        distance[destination] = 0;
        while (!q.isEmpty()) {
            int cur = q.poll();
            
            for (int nxt : maps[cur]) {
                if (distance[nxt] ==-1) {
                    distance[nxt] = distance[cur] +1;
                    q.offer(nxt);
                }
            }
        }
        
        int [] answer = new int[sources.length];
        int idx= 0;
        for (int s : sources) {
            answer[idx] = distance[s];
            idx+=1;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque
def solution(n, roads, sources, destination):
    answer = []
    
#     graph = [[] *(n+1) for _ in range (n+1)]
 

#     # graph 양방향 만들기 
#     for road in roads :
#         start, end = road[0], road[1]
#         graph[start].append (end)
#         graph[end].append (start)
    
#     # bfs로직 -> python에서는 자동으로 None 타입을 반환함
#     def bfs (start) :
#         distance = [-1] *(n+1)
#         distance[start] = 0
        
#         q= deque ([start])
        
#         while q :
#             cur = q.popleft ()
            
#             for nxt in graph[cur] :
#                 if distance[nxt] ==-1 :
#                     distance[nxt] = distance[cur] +1
#                     q.append(nxt)
#         return distance
#   # destination -> source 한번만 bfs조지면 됨

#     res = bfs (destination)
    
#     for start_node in sources:
#         answer.append(res[start_node])
    
    
    
    
    
    
    
    
    
    
    graph = [[] for _ in range (n+1)]
    
    for r in roads :
        start = r[0]
        end = r[1]
        
        graph[start].append(end)
        graph[end].append(start)
        
    distance = [-1]*(n+1)
    
    def bfs(start): 
        distance[start] = 0
        q= deque([start])
        while q :
            cur_node = q.popleft()
            for nxt_node in graph[cur_node] :
                if distance[nxt_node] == -1 :
                    distance[nxt_node] = distance[cur_node] +1
                    q.append(nxt_node)
        return 
    
    bfs(destination)
    
    for target_source in sources :
        answer.append (distance[target_source]) 
            
             
    return answer